test/faq.cpp

```diff
@@ -0,0 +1,184 @@
+    }
+    std::cout << std::endl;
+  }
+  std::cout << std::endl;
+  // Output:
+  //   0 1 2
+  //   3 4 5
+  //   6 7 8
+  //   9 10 11
+
+  a.reshape({2, 6});
+  for (auto y : a.y()) {
+    for (auto x : a.x()) {
+      std::cout << a(x, y) << " ";
+    }
+    std::cout << std::endl;
+  }
+  std::cout << std::endl;
+  // Output:
+  //   0 1
+  //   3 4
+  //   6 7
+  //   9 10
+  //   0 0
+  //   0 0
+  // Observe that the right column of the original array has been lost, and
+  // two default-constructed rows have been added to the bottom of the array.
+
+  // A: We can also reinterpret the shape of an existing array using
+  // `array::set_shape`:
+  a.set_shape({4, 3});
+  for (auto y : a.y()) {
+    for (auto x : a.x()) {
+      std::cout << a(x, y) << " ";
+    }
+    std::cout << std::endl;
+  }
+  std::cout << std::endl;
+  // Output:
+  //   0 1 3 4
+  //   6 7 9 10
+  //   0 0 0 0
+  // Observe that this has not removed or added any values from the array, the
+  // underlying memory has simply be reinterpreted as a different array.
+
+  // A: We can also use `reinterpret_shape` to make a new `array_ref` with the
+  // new shape:
+  auto a_reshaped = reinterpret_shape(a, dense_shape<2>{4, 3});
+  ASSERT(a_reshaped == a);
+
+  // Q: `array`'s move constructor requires the source array to have the same
+  // shape type. How do I move ownership to an array of a different shape type?
+
+  // A: The helper function `move_reinterpret_shape` combines move construction
+  // with `reinterpret_shape`:
+  array_of_rank<int, 3> source({3, 4, 5});
+  dense_array<int, 3> dest = move_reinterpret_shape<dense_shape<3>>(std::move(source));
+
+  // This can fail at runtime if the source shape is not compatible with the
+  // destination shape.
+}
+
+TEST(faq_crop) {
+  // Q: Cropping in this library is weird. How do I crop an array the way
+  // (my favorite library) does it?
+
+  // A: After cropping, the resulting array will have a min corresponding
+  // to the cropped region:
+  dense_array<int, 1> array({100});
+  for (int i = 0; i < array.size(); i++) {
+    array(i) = i;
+  }
+  const int crop_begin = 25;
+  const int crop_end = 50;
+  auto cropped = array(r(crop_begin, crop_end));
+  for (int i = crop_begin; i < crop_end; i++) {
+    ASSERT_EQ(array(i), cropped(i));
+  }
+
+  // This differs from most alternative libraries. To match this behavior,
+  // the `min` of the resulting cropped array needs to be changed to 0:
+  cropped.shape().dim<0>().set_min(0);
+  for (int i = crop_begin; i < crop_end; i++) {
+    ASSERT_EQ(array(i), cropped(i - crop_begin));
+  }
+
+  // The reason array works this way is to enable transparent tiling
+  // optimizations of algorithms.
+}
+
+TEST(faq_stack_allocation) {
+  // Q: How do I allocate an array on the stack?
+
+  // A: Use `auto_allocator<>` as the allocator for your arrays.
+
+  // Define an allocator that has storage for up to 100 elements.
+  using StackAllocator = auto_allocator<int, 100>;
```

Comment:
Oof, it took me a few minutes to wrap my head around this. `std::allocator` is stateless, but `auto_allocator` is not - it owns a buffer. I was trying to figure out where it's allocated - only to remember that it's a template argument - so the instance actually owns the allocator instance!

Not sure how to point this out (or if that's necessary. If you understand allocators then it's not a problem, and the documentation assumes you're pretty good at C++ :)).

Reply:
Yes, this is something that needs more documentation, in the array.h header actually. The docs on these allocators should explain they can't actually be used with STL containers, because they're stateful. 

It's actually really annoying to me that STL allocators are stateful because they preclude exactly this use case, which is very powerful. Replicating `absl::InlinedVector` goes from just a simple allocator like this one to basically a total rewrite of `std::vector` :(

Reply:
It's actually even more annoying than I remembered: STL allocators have all this machinery that looks like it *attempts* to enable stateful allocators, like these flags such as `propagate_on_container_assignment`. But no STL I tested respects those flags, argh!